common/blobstore/s3store/client.go

```diff
@@ -0,0 +1,321 @@
+	defer cancel()
+	tags, err := getTags(ctx, c.s3cli, bucket, key.String())
+	if err != nil {
+		return nil, err
+	}
+	return tags, nil
+}
+
+// Exists returns false when the bucket does not exist or the bucket and key does not exist
+func (c *client) Exists(ctx context.Context, bucket string, key blob.Key) (bool, error) {
+	ctx, cancel := c.ensureContextTimeout(ctx)
+	defer cancel()
+	_, err := c.s3cli.HeadObjectWithContext(ctx, &s3.HeadObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		aerr, ok := err.(awserr.Error)
+		if ok && (aerr.Code() == "NotFound") {
+			return false, nil
+		}
+
+		return false, err
+	}
+	return true, nil
+}
+
+// Delete will always return true whether or not the specific key exists
+// S3 DeleteObject gives no indication if the key exists
+func (c *client) Delete(ctx context.Context, bucket string, key blob.Key) (bool, error) {
+	ctx, cancel := c.ensureContextTimeout(ctx)
+	defer cancel()
+	_, err := c.s3cli.DeleteObjectWithContext(ctx, &s3.DeleteObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return false, blobstore.ErrBucketNotExists
+			}
+		}
+		return false, err
+
+	}
+	return true, nil
+}
+
+func (c *client) ListByPrefix(ctx context.Context, bucket string, prefix string) ([]blob.Key, error) {
+	ctx, cancel := c.ensureContextTimeout(ctx)
+	defer cancel()
+	results, err := c.s3cli.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, err
+	}
+	var keys = make([]blob.Key, len(results.Contents), len(results.Contents))
+	for i, v := range results.Contents {
+		key, err := blob.NewKeyFromString(*v.Key)
+		if err != nil {
+			return nil, blobstore.ErrConstructKey
+		}
+		keys[i] = key
+	}
+	return keys, nil
+}
+
+func (c *client) BucketMetadata(ctx context.Context, bucket string) (*blobstore.BucketMetadataResponse, error) {
+	ctx, cancel := c.ensureContextTimeout(ctx)
+	defer cancel()
+	results, err := c.s3cli.GetBucketAclWithContext(ctx, &s3.GetBucketAclInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, err
+	}
+
+	lifecycleResults, err := c.s3cli.GetBucketLifecycleConfigurationWithContext(ctx, &s3.GetBucketLifecycleConfigurationInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, err
+	}
+	retentionDays := 0
+	for _, v := range lifecycleResults.Rules {
```

Comment:
Are we iterating through these rules in chronological order? If so will retention days always be correct at the end of this loop? The way I am reading this code is we select the highest Expiration.Days that has ever been seen even if that is higher than the current Expiration.Days? Maybe I am just misunderstanding these APIs?